iroha_p2p/src/peer.rs

```diff
@@ -0,0 +1,325 @@
+    }
+}
+
+#[async_trait::async_trait]
+impl<E: Encryptor + Send + 'static> Handler<MessageResult> for Peer<E> {
+    type Result = ();
+
+    async fn handle(&mut self, msg: MessageResult) {
+        match msg {
+            MessageResult::Ok { message } => {
+                // We got handshake or the real message from other peer
+                match message {
+                    Message::Data { data } => {
+                        let data = match &self.cipher {
+                            None => { data }
+                            Some(cipher) => {
+                                match cipher.decrypt_easy(&DEFAULT_AAD[..], data.as_slice()) {
+                                    Ok(data) => data,
+                                    Err(e) => {
+                                        warn!("Error decrypting message! {}", e);
+                                        self.state = State::Error;
+                                        return;
+                                    }
+                                }
+                            }
+                        };
+                        let msg = ReceivedPeerMsg { data, id: self.id.clone() };
+                        self.network_addr.do_send(msg).await;
+                    },
+                    Message::Error => {
+                        self.state = State::Error;
+                    }
+                }
+            }
+            MessageResult::Err { error } => {
+                warn!("Error in peer read! {}", error);
+            }
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl<E: Encryptor + Send + 'static> Handler<SendPeerMsg> for Peer<E> {
+    type Result = ();
+
+    async fn handle(&mut self, msg: SendPeerMsg) {
+        if self.write.is_none() {
+            warn!("Cannot send message to peer, as we are not connected!");
+            return;
+        }
+
+        let data = match &self.cipher {
+            None => msg.data,
+            Some(cipher) => {
+                match cipher.encrypt_easy(&DEFAULT_AAD[..], msg.data.as_slice()) {
+                    Ok(data) => data,
+                    Err(e) => {
+                        warn!("Error encrypting message! {}", e);
+                        self.state = State::Error;
+                        return;
+                    }
+                }
+            }
+        };
+        let write_half = self.write.as_mut();
+        if let Err(e) = send_message(&mut write_half.unwrap(), data.as_slice()).await {
+            warn!("Error sending message to peer! {}", e);
+            self.state = State::Error;
+        }
+    }
+}
+
+#[derive(Copy, Clone, PartialEq, Eq)]
+pub enum State {
+    Connecting,
+    ConnectedTo,
+    ConnectedFrom,
+    Ready,
+    Error
+}
+
+pub async fn readable<T: AsRef<TcpStream>>(stream: T) -> io::Result<()> {
+    stream.as_ref().readable().await
+}
+
+pub async fn writable<T: AsRef<TcpStream>>(stream: T) -> io::Result<()> {
+    stream.as_ref().writable().await
+}
```

Comment:
You use those functions only once. Also they dont seem to abstract over much of a logic. Could you just call them directly on stream halves?

Reply:
Removed!